src/fifelse.c

```diff
@@ -343,3 +343,193 @@ SEXP fcaseR(SEXP na, SEXP rho, SEXP args) {
   UNPROTECT(nprotect);
   return ans;
 }
+
+SEXP topnR(SEXP vec, SEXP n, SEXP dec) {
+	int nprotect = 0;
+	int64_t i, j, idx = 0;
+	const int64_t len0 = asInteger(n);
+	const int64_t len1 = xlength(vec);
+	
+	if (isS4(vec) && !INHERITS(vec, char_nanotime)) {
+		error("S4 class objects (excluding nanotime) are not supported.");
+	}
+	if (len0 > 1000) {
+		error("Function 'topn' is not built for large value of 'n'. The algorithm is made for small values. Please prefer the 'order' if you want to proceed with such large value.");
+	}
+	if (len0 > len1) {
+		error("'n' cannot be larger than length of 'vec'.");
```

Comment:
A very likely use case is ```topn``` by group. I think most people familiar with ```top_n``` would expect this to work:
```
set.seed(123L)
dt = data.table(x = rnorm(10), grp = sample(5L, 10L, replace = T))
dt[dt[, .I[topn(x, 2L)], by = grp]$V1, ]

##Error: 'n' cannot be larger than length of 'vec'.
```

Reply:
so you suggest we return something like below?

```R
> dt[dt[, .I[order(x)[1:2]], by = grp]$V1, ]
             x grp
 1: -0.6868529   1
 2: -0.5604756   1
 3: -0.2301775   4
 4:         NA  NA
 5:  0.1292877   5
 6:         NA  NA
 7: -0.4456620   3
 8:  1.7150650   3
 9: -1.2650612   2
10:  0.4609162   2
```

Reply:
My preference would it be to return ```seq_len(n)``` and maybe have a warning that there are less records than ```n```.

```
dt[dt[, .I[order(x)[1:min(.N,2)]], by = grp]$V1, ]

            x   grp
        <num> <int>
1: -0.6868529     1
2: -0.5604756     1
3: -0.2301775     4
4:  0.1292877     5
5: -0.4456620     3
6:  1.7150650     3
7: -1.2650612     2
8:  0.4609162     2
```

P.S., this PR is great. 
```
set.seed(123)
dt = data.table(x = rnorm(1e7), grp = sample(100, 1e7, TRUE))

microbenchmark::microbenchmark(
dt[, .I[topn(x, 5L)], by = grp],
dt[order(-x), .I[seq_len(5L)], by = grp],
dt[, .I[frank(-x, ties.method = "min") <= 5L], by = grp], ##closest to dplyr::top_n
times = 5L
)

Unit: milliseconds
                                                     expr       min        lq      mean    median        uq       max neval
                          dt[, .I[topn(x, 5L)], by = grp]  157.9859  158.9587  161.4327  160.2456  160.6012  169.3721     5
                 dt[order(-x), .I[seq_len(5L)], by = grp] 1336.8226 1343.3002 1347.5568 1345.3929 1350.1008 1362.1676     5
 dt[, .I[frank(-x, ties.method = "min") <= 5L], by = grp] 1432.7179 1459.2029 1529.9473 1498.1096 1626.5848 1633.1211     5
```

Reply:
It should be now working like you asked.

Reply:
Thanks! 

This is more for future thought. I was also going to suggest the maintainers look at updating the [Database-like ops benchmark](https://h2oai.github.io/db-benchmark/) for top n by group as ```dt[, .(largest2_v3 = v3[topn(v3, 2L, decreasing = TRUE)], by = id6]```. But as the number of groups increase, there seems to be some overhead. I assume this is a memory and ```gforce``` type of issue. I can log a separate issue once this is merged for better grouping performance. 

<details>
  <summary>Details, benchmark, and verbose message</summary>


There are two rows with ******* which identify the main performance different - for ```topn``` it is ```collecting discontiguous groups took 0.616s for 10000 groups``` versus only 0.016s for the other method.

``` r

##remotes::install_github("Rdatatable/data.table", ref = "topn")

library(data.table)
N = 1000000L; K = 100L; set.seed(108L)
DT = data.table(id6 = sample(N/K, N, TRUE), v3 = runif(N, max = 100))

identical(DT[DT[order(-v3), head(.I, 2L), by = id6]$V1, .(id6, largest2_v3 = v3)][order(-largest2_v3, id6)],
          DT[, .(largest2_v3 = v3[topn(v3, 2L, decreasing = TRUE)]), by = id6][order(-largest2_v3, id6)])
#> [1] TRUE

microbenchmark::microbenchmark(
  DT[DT[order(-v3), head(.I, 2L), by = id6]$V1, .(id6, largest2_v3 = v3)],
  DT[, .(largest2_v3 = v3[topn(v3, 2L, decreasing = TRUE)]), by = id6]
  , times = 10L
  )
#> Unit: milliseconds
#>                                                                     expr
#>  DT[DT[order(-v3), head(.I, 2L), by = id6]$V1, .(id6, largest2_v3 = v3)]
#>     DT[, .(largest2_v3 = v3[topn(v3, 2L, decreasing = TRUE)]), by = id6]
#>       min       lq     mean   median       uq      max neval
#>  242.2958 248.2904 254.6175 250.4815 255.1830 288.5708    10
#>  674.9867 680.0911 722.0210 682.1045 697.4592 915.3662    10

DT[DT[order(-v3), head(.I, 2L), by = id6, verbose = T]$V1, .(id6, largest2_v3 = v3)]
#> forder.c received 1000000 rows and 1 columns
#> i clause present and columns used in by detected, only these subset: id6 
#> Detected that j uses these columns: <none> 
#> Finding groups using forderv ... forder.c received 1000000 rows and 1 columns
#> 0.000s elapsed (0.000s cpu) 
#> Finding group sizes from the positions (can be avoided to save RAM) ... 0.000s elapsed (0.000s cpu) 
#> Getting back original order ... forder.c received a vector type 'integer' length 10000
#> 0.000s elapsed (0.000s cpu) 
#> lapply optimization is on, j unchanged as 'head(.I, 2L)'
#> GForce is on, left j unchanged
#> Old mean optimization is on, left j unchanged.
#> Making each group and running j (GForce FALSE) ... 
#>   collecting discontiguous groups took 0.016s for 10000 groups *********
#>   eval(j) took 0.090s for 10000 calls
#> 0.110s elapsed (0.110s cpu)

DT[, .(largest2_v3 = v3[topn(v3, 2L, decreasing = TRUE)]), by = id6, verbose = TRUE]
#> Detected that j uses these columns: v3 
#> Finding groups using forderv ... forder.c received 1000000 rows and 1 columns
#> 0.000s elapsed (0.000s cpu) 
#> Finding group sizes from the positions (can be avoided to save RAM) ... 0.000s elapsed (0.000s cpu) 
#> Getting back original order ... forder.c received a vector type 'integer' length 10000
#> 0.000s elapsed (0.000s cpu) 
#> lapply optimization is on, j unchanged as 'list(v3[topn(v3, 2L, decreasing = TRUE)])'
#> GForce is on, left j unchanged
#> Old mean optimization is on, left j unchanged.
#> Making each group and running j (GForce FALSE) ... 
#>   collecting discontiguous groups took 0.616s for 10000 groups  **********
#>   eval(j) took 0.034s for 10000 calls
#> 0.690s elapsed (0.440s cpu)

```
</details>